src/single-spa-react.ts

```diff
@@ -0,0 +1,66 @@
+import {
+  chooseDomElementGetter,
+  type DomElementGetterOpts,
+} from "dom-element-getter-helpers";
+import type { ReactNode } from "react";
+import type {
+  createRoot,
+  hydrateRoot,
+  RootOptions,
+  Root,
+} from "react-dom/client";
+
+export interface SingleSpaReactOpts extends DomElementGetterOpts {
+  createRoot: typeof createRoot;
+  hydrateRoot: typeof hydrateRoot;
```

Comment:
Slightly more complicated, but something like this:

```ts
type AtLeastOne<T> = { [Key in keyof T]: Pick<T, Key> }[keyof T];

export type SingleSpaReactOpts = DomElementGetterOpts & AtLeastOne<{
  createRoot: typeof createRoot;
  hydrateRoot: typeof hydrateRoot;
}> & {
  rootOptions?: RootOptions;
  createReactNode<T = any>(props: T): ReactNode | Promise<ReactNode>
}
```

Actually matches the requirements on line 27 and makes it possible (in TS) to pass in either `createRoot()` or `hydrateRoot()` or both.

Reply:
For the "both" case, it may be nice to have something like the old `renderType` property since that allows isomorphic app definitions, but it's probably not required.